### PR TITLE
[API] Add stream RPC for `RateLimitService` and add the flag to let controller specifies whether unary or stream RPC will be used

### PR DESCRIPTION
**Pull request by tyxia**
- Introduce bi-direction stream gRPC communication to RLS sever
- Add the flag to make the stream vs unary communication configurable 

Signed-off-by: Tianyu Xia <tyxia@google.com>

Risk Level: LOW
Testing: CI
Release Notes: Not implemented

### CODE REVIEW

File: api/envoy/config/ratelimit/v3/rls.proto [hunk 1]
```diff
@@ -31,4 +32,8 @@ message RateLimitServiceConfig {
   // API version for rate limit transport protocol. This describes the rate limit gRPC endpoint and
   // version of messages used on the wire.
   core.v3.ApiVersion transport_api_version = 4 [(validate.rules).enum = {defined_only: true}];
+
+  // Specifies whether the RPC communication with RLS server should be unary or streaming RPC.
```
Comment (adisuissa) [inline api/envoy/config/ratelimit/v3/rls.proto:36]: Add default value to the comment.
If the default value might be changed in the future, then you should consider the Bool WKT.

File: api/envoy/service/ratelimit/v3/rls.proto [hunk 2]
```diff
@@ -25,6 +25,11 @@ service RateLimitService {
   // Determine whether rate limiting should take place.
   rpc ShouldRateLimit(RateLimitRequest) returns (RateLimitResponse) {
   }
+
+  // Determine whether rate limiting should take place (Stream RPC version).
```
Comment (adisuissa) [inline api/envoy/service/ratelimit/v3/rls.proto:29]: nit:
```suggestion
  // Determine whether rate limiting should take place (Bidi-Stream RPC version).
```

**Review by adisuissa — COMMENTED** [review]
Overall LGTM.
Seems to be a very small change. Consider just adding this to the PR with the code-changes.

File: api/envoy/config/ratelimit/v3/rls.proto (hunk 1 above)
Comment (tyxia) [reply api/envoy/config/ratelimit/v3/rls.proto]: The bool default value in proto is false(https://developers.google.com/protocol-buffers/docs/proto3#default), which is intention here. 
I guess you mean add this info to comment? Done.

File: api/envoy/service/ratelimit/v3/rls.proto (hunk 2 above)
Comment (tyxia) [reply api/envoy/service/ratelimit/v3/rls.proto]: Sure. Done

**Review by tyxia — COMMENTED** [review]
Thanks for the review!

**Review by adisuissa — APPROVED** [review]
/lgtm api

File: api/envoy/service/ratelimit/v3/rls.proto [hunk 3]
```diff
@@ -25,6 +25,11 @@ service RateLimitService {
   // Determine whether rate limiting should take place.
   rpc ShouldRateLimit(RateLimitRequest) returns (RateLimitResponse) {
   }
+
+  // Determine whether rate limiting should take place (Bidi-Stream RPC version).
```
Comment (markdroth) [inline api/envoy/service/ratelimit/v3/rls.proto:29]: What are the semantics of this bidi stream?  For example, can multiple requests be in flight at the same time on the same stream?  If so, does the protocol provide enough information for the client to know which response is associated with which request?

Comment (tyxia) [reply api/envoy/service/ratelimit/v3/rls.proto:29]: Good questions! I have some initial implementation.

> can multiple requests be in flight at the same time on the same stream? 

More specially, for sending message, either `GoogleAsyncStreamImpl::sendMessageRaw()` or `AsyncStreamImpl::sendMessage()` will be invoked based on use cases. The send is invoked [here](https://github.com/envoyproxy/envoy/blob/a5927dc3f59e79772a6007b3b5538e86b2a597b0/source/common/grpc/typed_async_client.cc#L15). 

AFAIU, only one request (one `envoy::service::ratelimit::v3::RateLimitRequest`, serialized message) will be sent at one time, but multiple requests (not at the same time) can be sent as long as stream is opened. 

Does this answer your questions? Also, I would like to hear your thoughts on this. Thanks!

**Pull request closed without merging**